oracle-token/src/lib.rs

```diff
@@ -0,0 +1,1170 @@
+    }
+}
+
+impl<T> DecodeError<T> for Error {
+    fn type_of() -> &'static str {
+        "Oracle Token Error"
+    }
+}
+impl PrintProgramError for Error {
+    fn print<E>(&self)
+    where
+        E: 'static + std::error::Error + DecodeError<E> + PrintProgramError + FromPrimitive,
+    {
+        match self {
+            Error::AlreadyInUse => info!("Error: AlreadyInUse"),
+            Error::InvalidTokenState => info!("Error: InvalidTokenState"),
+            Error::InvalidState => info!("Error: InvalidState"),
+            Error::InvalidInput => info!("Error: InvalidInput"),
+            Error::CalculationFailure => info!("Error: CalculationFailure"),
+        }
+    }
+}
+
+pub type Ratio = (u64, u64);
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct OracleValues {
+    latest: u64,
+    values: [(Slot, Ratio); 32],
+}
+impl OracleValues {
+    fn from_account_info(_info: &AccountInfo) -> Result<Self, ProgramError> {
+        unimplemented!();
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct OracleToken {
+    self_pubkey: Pubkey,
+    authority: Pubkey,
+    token_program: Pubkey,
+    oracle_values: Pubkey,
+    colleteral: Pubkey,
+    pool_mint: Pubkey,
+    synthetic_mint: Pubkey,
+    pool_supply: u64,
+    synthetic_supply: u64,
+    info: OracleTokenInfo,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct Deposit {
+    slot: Slot,
+    amount: u64,
+    user: Pubkey,
+}
+
+impl Deposit {
+    fn calculate(
+        &self,
+        clock: &Clock,
+        info: &OracleTokenInfo,
+        pool: u64,
+        colleteral: u64,
+        synthetic: u64,
+        oracle_values: &OracleValues,
+    ) -> Option<u64> {
+        let (slot, price) = oracle_values.values.get(oracle_values.latest as usize)?;
+        if *slot < self.slot + info.min_slots && clock.slot < self.slot + info.max_slots {
+            return None;
+        }
+        let obligation = synthetic as f64 * price.0 as f64 / price.1 as f64;
+        let base = colleteral as f64 - obligation;
+        let percentage = self.amount as f64 / base;
+        let amount = percentage * pool as f64;
+        Some(amount as u64)
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct Withdraw {
+    slot: Slot,
+    amount: u64,
+    user: Pubkey,
+}
+
+impl Withdraw {
+    fn calculate(
+        &self,
+        clock: &Clock,
+        info: &OracleTokenInfo,
+        pool: u64,
+        colleteral: u64,
+        synthetic: u64,
+        oracle_values: &OracleValues,
+    ) -> Option<u64> {
+        let (slot, price) = oracle_values.values.get(oracle_values.latest as usize)?;
+        if *slot < self.slot + info.min_slots && clock.slot < self.slot + info.max_slots {
+            return None;
+        }
+        let obligation = synthetic as f64 * price.0 as f64 / price.1 as f64;
+        let percentage = self.amount as f64 / pool as f64;
```

Comment:
Assuming that there are 100 pool tokens and 100 collateral but no synths.
If I withdraw 100 pool tokens they are burned in `init_withdraw` which means that here the pool supply will be 0 which would result in an error.
The calculation would also yield wrong results when withdrawing 99 pool tokens etc.

Solution: use `pool + self.amount`